"WiFi Unknown connection status 0" in empty ESP32

### The problem

Recently I'm experiencing the same WiFi issue with 3 different ESP32 boards. The yaml file has only the base configuration, with no added components, just the plain suggested config.

I'm flashing using manual download, and just after the flashing is completed it tries to connect to the network, gets "Auth expired" and "Association Expired" messages to finally collapse repeating "WiFi Unknown connection status 0" ad eternum. 

I've checked the [#2742](https://github.com/esphome/issues/issues/2742) issue and even using the password without secrets it crashes. I also went trough [#455](https://github.com/esphome/issues/issues/455) issue and it seems it's not the problem either, because my config is almost blank.

I've tried several fixes like fixed ip address, disabling logger or removing components, but nothing seems to work. I also removed wifi, captive_portal, web:server, ota and api, installed over download and the boards are working fine without trying to communicate.

Also I have a couple more ESP32 boards arround working in different automation projects with successful EspHome updates (i.e. sprinklers, alarm, etc.). 

I've ran out of new boards, so in the meanwhile any help would be preciated here.

### Which version of ESPHome has the issue?

2021.12.3

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

core-2022.3.7

### What platform are you using?

ESP32

### Board

_No response_

### Component causing the issue

MAybe WiFi?

### Example YAML snippet

```yaml
wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "Prueba GPIO"
    password: !secret fallback_password
    
captive_portal:
logger:
api:
ota:

esphome:
  name: ${nombre}
  platform: ESP32
  board: esp-wrover-kit
  
substitutions:
  
  # Nombre del dispositivo en Home Assistant
  nombre: 'prueba-gpio'
```


### Anything in the logs that might be useful for us?

```txt
Using 'COM4' as serial port.
Writing at 0x0008c000... (100 %)Wrote 889008 bytes (509440 compressed) at 0x00010000 in 12.1 seconds (effective 588.2 kbit/s)...
Hash of data verified.

Leaving...
Hard Resetting...
Done! Flashing is complete!

Showing logs:
[21:11:45][I][logger:214]: Log initialized
[21:11:45][C][ota:458]: There have been 0 suspected unsuccessful boot attempts.
[21:11:45][D][esp32.preferences:114]: Saving preferences to flash...
[21:11:45][I][app:029]: Running through setup()...
[21:11:45][C][wifi:037]: Setting up WiFi...
[21:11:46][D][wifi:370]: Starting scan...
[21:11:47][D][wifi:385]: Found networks:
[21:11:47][I][wifi:429]: - 'VTR-6000486' (18:35:D1:B1:3F:D9) ▂▄▆█
[21:11:47][D][wifi:430]:     Channel: 11
[21:11:47][D][wifi:431]:     RSSI: -60 dB
[21:11:47][D][wifi:434]: - 'DIRECT-CBZEUSmsHo' (26:DB:C9:4B:53:5A) ▂▄▆█
[21:11:47][I][wifi:245]: WiFi Connecting to 'VTR-6000486'...
[21:11:49][W][wifi_esp32:495]: Event: Disconnected ssid='VTR-6000486' bssid=18:35:D1:B1:3F:D9 reason='Auth Expired'
[21:11:49][W][wifi:533]: Error while connecting to network.
[21:11:49][W][wifi:568]: Restarting WiFi adapter...
[21:11:49][I][wifi:245]: WiFi Connecting to 'VTR-6000486'...
[21:11:50][W][wifi_esp32:495]: Event: Disconnected ssid='VTR-6000486' bssid=18:35:D1:B1:3F:D9 reason='Association Expired'
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:50][W][wifi:554]: WiFi Unknown connection status 0
[21:11:51][W][wifi:554]: WiFi Unknown connection status 0
```


### Additional information

EDIT: This morning, came to the computer and ESP32 was up and running. Then I tried to check the logs via EspHome Flasher and it crashed again. It seems to oscillate between "Wifi Unknown connection status" and connected back and forth several times per minute.

Looking back through the output from esphome wizard I find some messages which might be significant...
```
Library Manager: Installing Hash
Library Manager: Warning! Could not install dependency {'name': 'Hash', 'platforms': ['espressif8266']} for package 'ESPAsyncWebServer-esphome'
Library Manager: Installing ESP8266WiFi
Library Manager: Warning! Could not install dependency {'name': 'ESP8266WiFi', 'platforms': ['espressif8266']} for package 'ESPAsyncWebServer-esphome'
Dependency Graph
```
and later
```
Compiling .pioenvs/esp32-cam/FrameworkArduino/esp32-hal-touch.c.o
/home/don/.platformio/packages/framework-arduinoespressif32/cores/esp32/esp32-hal-spi.c: In function 'spiTransferBytesNL':
/home/don/.platformio/packages/framework-arduinoespressif32/cores/esp32/esp32-hal-spi.c:922:39: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
                 uint8_t * last_out8 = &result[c_longs-1];
                                       ^
/home/don/.platformio/packages/framework-arduinoespressif32/cores/esp32/esp32-hal-spi.c:923:40: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
                 uint8_t * last_data8 = &last_data;
                                        ^
Compiling .pioenvs/esp32-cam/FrameworkArduino/esp32-hal-uart.c.o
```

--- a/esphome/components/wifi/wifi_component.cpp
+++ b/esphome/components/wifi/wifi_component.cpp
@@ -555,6 +555,7 @@ void WiFiComponent::check_connecting_finished() {
   }
 
   ESP_LOGW(TAG, "WiFi Unknown connection status %d", (int) status);
+  this->retry_connect();
 }
 
 void WiFiComponent::retry_connect() {

